lib/websocket_proxy.rb

```diff
@@ -0,0 +1,74 @@
+class WebsocketProxy
+  attr_reader :env, :url, :error
+
+  def initialize(env, console)
+    @env = env
+    @id = SecureRandom.uuid
+    @console = console
+
+    secure = Rack::Request.new(env).ssl?
+    scheme = secure ? 'wss:' : 'ws:'
+    @url = scheme + '//' + env['HTTP_HOST'] + env['REQUEST_URI']
+    @driver = WebSocket::Driver.rack(self, :protocols => %w(binary))
+
+    begin
+      # Hijack the socket from the Rack middleware
+      @ws = env['rack.hijack'].call
+      # Set up the socket client for the proxy
+      @sock = TCPSocket.open(@console.host_name, @console.port)
+      init_ssl if @console.ssl
+    rescue
+      @error = true
+    end
+
+    @driver.on(:open) { @console.update(:opened => true) }
+
+    @driver.on(:message) do |msg|
+      @ssl ? @ssl.syswrite(msg.data.pack('C*')) : @sock.write(msg.data.pack('C*'))
+    end
+
+    @driver.on(:close) { cleanup }
+
+    @driver.start
+  end
+
+  def transmit(sockets, is_ws)
+    # Do not read when the other end is not ready for writing
+    return unless is_ws ? sockets.include?(@ws) : sockets.include?(@sock)
+
+    if is_ws
+      data = @ws.recv_nonblock(65_535)
+      @driver.parse(data)
+    else
+      data = @ssl ? @ssl.sysread(65_535) : @sock.recv_nonblock(65_535)
+      @driver.binary(data)
+    end
+  end
+
+  def cleanup
+    @console.destroy unless @console.destroyed?
+    @sock.close unless @sock.closed?
+    @ws.close unless @ws.closed?
+  end
+
+  def descriptors
+    [@ws, @sock]
+  end
+
+  def write(string)
+    @ws.write(string)
+  end
+
+  private
+
+  def init_ssl
+    context = OpenSSL::SSL::SSLContext.new
+    context.cert = OpenSSL::X509::Certificate.new(File.open('certs/server.cer'))
+    context.key = OpenSSL::PKey::RSA.new(File.open('certs/server.cer.key'))
+    context.ssl_version = :SSLv23
+    context.verify_depth = OpenSSL::SSL::VERIFY_NONE
+    @ssl = OpenSSL::SSL::SSLSocket.new(@sock, context)
+    @ssl.sync_close = true
+    @ssl.connect
```

Comment:
So the `cert` and `key` say that we trust our own server cert.
So when it connects to our server, it is valid.
These all look good.

`verify_depth=VERIFY_NONE` - says that it will be content connecting to any hostname (and ignore possible man in the middle attacks). We are connecting to `localhost` so while this is not an issue, the security community is gun shy about this one. Is there a way we can do `VERIFY_PEER` or some other value?